### ISSUE TITLE
possible to pass a function to a macro?  or find the identifier for the current function?

### ISSUE THREAD

**Issue by retrac0** [KarolS/millfork#118]
Hello again!  I am sorry I haven't been active much.   I have been rewriting part of my Atari 2600 demo/monstrosity in Millfork and it's been pretty great so far.  I've run into a minor hitch with some routines to handle bankswitching.  In my case, bankswitching involves one of four windows, so I extract the window from the target's address hi byte, along with the bank segment constant:

```
macro void swap (void a) { 
    bankswitch_reg = a.segment | ((hi(a.addr) & %0000_1100) << 4)
}

segment(somesegment) word x;

void main () {
    swap(x)
}
```

With optimization this compiles down to a LDA# STA sequence (perfect!) and it works great for variables.  But if I try to pass swap() a function I get a type error about "`myfunc` is not a TypedThing" which I assume means it's not really typed void and so cannot be passed like that.  Of course there is a pretty simple workaround by extracting .segment and .addr and passing them to a macro which just takes them as arguments:

```
swapfn(myfunc.segment, myfunc.addr) 
```

But I was wondering if there's a cleaner way so I can use the first approach?

Also, while maybe a little neurotic, would it be possible determine the current function's address (or identifier to use with .addr) so I can do a compile-time check in swap() to ensure that I am not swapping out the code which is actively running?

With thanks.

**Comment by KarolS** [comment]
There is currently no way to make it cleaner, but you found a really nice place for potential improvements, thanks!

So there are two potentials for improvement here: 

* functions are not "typed things", which is the internal Millfork term for variables, constants and similar things – functions are merely "things" – and right now, macro parameters have to be "typed things", even if we don't care about their types

* macros should preferably know the function and segment they're in – and even normal functions should be able to refer to themselves without repeating their name

I'll figure out some improvements here, they should land in the next release.

BTW, A2600 support in Millfork was more of an afterthought, in the lines of "it'll compile, so why not". It's surprising, but nice to hear that it turned out to be practical.

**Comment by retrac0** [comment]
Thanks!  It'll be a help.  I did notice the VCS support was minimalist, and wrote my own platform file and includes.   But MFK seems pretty general purpose and can fit into many of the VCS bankswitching arrangements.  And MFK is very fast... I've only needed to write about a dozen lines of the graphics kernel core loop in ASM.  Otherwise pure MFK.  

It is limited to the 128 B of RAM for regular variables.  VCS cartridges use a complicated system for cartridge RAM, where separate addresses are used for read/write (and RMW ops like inc/dec don't work).  I've had luck wrapping that in a few macros and using pointers.  With care it compiles to code very similar to what I would write in assembly. :)  I'll try to document some of it when I have time.
